autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java

```diff
@@ -0,0 +1,180 @@
+        }
+    }
+
+    protected void handleScaleUp(String hostGroupInstanceType, NewNodeManagerCandidates newNMCandidates, Integer existingHostGroupSize) {
+        Integer yarnRecommendedHostGroupCount =
+                newNMCandidates.getCandidates().stream()
+                        .filter(candidate -> candidate.getModelName().equalsIgnoreCase(hostGroupInstanceType))
+                        .findFirst().map(NewNodeManagerCandidates.Candidate::getCount)
+                        .orElseThrow(() ->
+                                new RuntimeException(String.format("Yarn Scaling API Response does not contain recommended host count " +
+                                                "for hostGroupInstanceType %s in Cluster %s, Yarn Response %s",
+                                        hostGroupInstanceType, cluster.getStackCrn(), newNMCandidates)));
+
+        int maxAllowedScaleUp = loadAlert.getLoadAlertConfiguration().getMaxResourceValue() - existingHostGroupSize;
+        maxAllowedScaleUp = maxAllowedScaleUp < 0 ? 0 : maxAllowedScaleUp;
+        int scaleUpCount = Math.min(maxAllowedScaleUp, yarnRecommendedHostGroupCount);
+
+        LOGGER.info("ScaleUp NodeCount {} for Cluster {}, HostGroup {]", scaleUpCount,
+                cluster.getStackCrn(), loadAlert.getScalingPolicy().getHostGroup());
+
+        if (scaleUpCount > 0) {
+            loadAlert.getScalingPolicy().setScalingAdjustment(scaleUpCount);
+            eventPublisher.publishEvent(new ScalingEvent(loadAlert));
+        }
+    }
+
+    protected void handleScaleDown(List<DecommissionCandidate> decommissionCandidates, Map<String, String> hostGroupFqdnsToInstanceId) {
+        Set<String> hostGroupFqdns = hostGroupFqdnsToInstanceId.keySet();
+        int maxAllowedScaleDown = hostGroupFqdns.size() - loadAlert.getLoadAlertConfiguration().getMinResourceValue();
+        maxAllowedScaleDown = maxAllowedScaleDown < 0 ? 0 : maxAllowedScaleDown;
+
+        List<String> decommissionHostGroupNodeIds = decommissionCandidates.stream()
+                .sorted(Comparator.comparingInt(DecommissionCandidate::getAmCount))
+                .map(DecommissionCandidate::getNodeId)
+                .map(nodeFqdn -> nodeFqdn.split(":")[0])
+                .filter(s -> hostGroupFqdns.contains(s))
+                .limit(maxAllowedScaleDown)
+                .map(nodeFqdn -> hostGroupFqdnsToInstanceId.get(nodeFqdn))
+                .collect(Collectors.toList());
+
+        LOGGER.info("ScaleDown NodeCount {} for Cluster {}, HostGroup {], NodeIds {}",
+                decommissionHostGroupNodeIds.size(), cluster.getStackCrn(), loadAlert.getScalingPolicy().getHostGroup(),
+                decommissionHostGroupNodeIds);
+
+        if (!decommissionHostGroupNodeIds.isEmpty()) {
+            loadAlert.setDecommissionNodeIds(decommissionHostGroupNodeIds);
+            eventPublisher.publishEvent(new ScalingEvent(loadAlert));
```

Comment:
However the event which gets published is changed - I think it'll be useful to include a timestamp in the event. That can be used to track metrics on how long from event generation to event processing - and generate alerts based on that. Will be useful for load testing as well.

Reply:
eventTimestamp is already present in base class.